### PR TITLE
FIX Ensure report columns can be extended

### PR DESCRIPTION
**Pull request by robbieaverill**
$summary_fields is not used when generating this report, which in itself it OK, but we need the columns to be extensible to the other modules can add to them (as well as third party devs)

### CODE REVIEW

File: src/Model/Package.php [hunk 1]
```diff
@@ -14,12 +14,6 @@ class Package extends DataObject
         'Type' => 'Varchar(255)',
     ];
 
-    private static $summary_fields = [
```
Comment (NightJar) [inline src/Model/Package.php]: By default all fields are summary fields, as I recall. I don't agree that `Type` `silverstripe-module` is relevant information for export (or plain text summaries anywhere else). Could you please revert this change?

File: src/Reports/SiteSummary.php [hunk 2]
```diff
@@ -55,16 +59,20 @@ public function columns()
...
     {
         Requirements::css('silverstripe-maintenance/css/sitesummary.css');
-        $grid = parent::getReportField();
-        $config = $grid->getConfig();
+
+        /** @var GridField $gridField */
+        $gridField = parent::getReportField();
+
+        /** @var GridFieldExportButton $exportButton */
+        $exportButton = $gridField->getConfig()->getComponentByType(GridFieldExportButton::class);
```
Comment (NightJar) [inline src/Reports/SiteSummary.php]: Although `getConfig` is relatively trivial, I think the way it was stored to a variable before was better in repeated use.
Could you please shift the `getComponentByType` call to use `$config` (after it's defined)?

File: src/Reports/SiteSummary.php [hunk 3]
```diff
@@ -55,16 +59,20 @@ public function columns()
...
+        $exportButton->setExportColumns(Package::create()->summaryFields());
```
Comment (NightJar) [inline src/Reports/SiteSummary.php]: `Package.summary_fields` config is used here.

File: src/Model/Package.php (hunk 1 above)
Comment (robbieaverill) [reply src/Model/Package.php]: Can you explain again why it's needed? When I checked the summary fields were not being used to generate the report

Comment (robbieaverill) [reply src/Model/Package.php]: There's a bug here. I'm going to fix it.

Comment (NightJar) [reply src/Model/Package.php]: My review comment summarises the notion too (in the conversation tab, not a diff comment like this thread), if that provides more context to my meaning :)

Comment (robbieaverill) [reply src/Model/Package.php]: I've pushed a commit which ensures that the same columns are used for exporting to CSV, which removes the confusion around this piece of code

File: src/Reports/SiteSummary.php [hunk 4]
```diff
@@ -55,16 +59,29 @@ public function columns()
...
+
+        /** @var GridFieldExportButton $exportButton */
+        $exportButton = $config->getComponentByType(GridFieldExportButton::class);
+
+        // Switch for an unformatted name and description over the HTML formatted summary field
+        $exportColumns = [
+            'Name' => 'Name',
+            'Description' => 'Description',
+        ] + $this->columns();
+        unset($exportColumns['Summary']);
```
Comment (NightJar) [inline src/Reports/SiteSummary.php]: I find this more confusing than the inbuilt `summaryFields` feature.

In conjunction with the previous two comments on this iteration of the PR I feel like leveraging pre-provided functionality to it's maximum is of more benefit.

Comments can always be added to make things clearer.

```
public function getReportField()
{
    $gridField = parent::getReportField();
    $config = $gridField->getConfig();
    $config->addComponents(
        // [...] refresh and link buttons
    );
    // The GridFieldExportButton uses our model's summary_fields by default
    // as opposed to this reports `columns` method above {@see Package::$summary_fields}
    return $gridField;
}
```

File: src/Model/Package.php (hunk 1 above)
Comment (NightJar) [reply src/Model/Package.php]: I still disagree with this approach. I think using the inbuilt functionality provided by SilverStripe for this purpose (`summary_fields` config) is more consistent overall and causes less cognitive friction when reading.

If anything I'd prefer that `summary_fields` here was used in the generation of the columns, having the modification for the HTML output happen where that's relevant - in the columns function on the report. The default output for a CSV export button is to use the `summary_fields` of the model, so that logic in `getReportField` is no longer needed and would actually result in both a reduction of code and confusion in this class - following the idiom 'simplest is best', which I think is what we're really striving for here.

Please see the other two comments.

File: src/Reports/SiteSummary.php [hunk 5]
```diff
@@ -41,10 +41,14 @@ public function sourceRecords($params)
...
     {
-        return [
+        $columns = [
             'Summary' => 'Description',
             'Version' => 'Version',
         ];
+
+        $this->extend('updateColumns', $columns);
+
+        return $columns;
```
Comment (NightJar) [inline src/Reports/SiteSummary.php]: Following on from previous comment (for `Package.php`)

I think I'm guilty of a lack of foresight, and that this is where the alterations to encapsulate the reports unique screen presentation should be.

```
public function columns()
{
    $columns = Package::create()->summaryFields();
    $unwantedColumns = array_flip(['Name', 'Description']);
    $columns = array_diff_key($columns, $unwantedColumns);
    $columns = array_merge(['Summary' => 'Description'], $columns);

    $this->extend('updateColumns', $columns);

    return $columns;
}
```

**Comment by robbieaverill** [comment]
Ok so I've adjusted this to use summary fields for both the CMS report view and the export view, but allow `SiteSummary::columns` to provide frontend only formatting decisions (such as making title and description into a single column called "summary").

How do you feel about this now?

**Comment by NightJar** [comment]
Looks great, clean and easy to read whilst still being specific about why things are being mutated. Nice :)
